src/typechecker.erl

```diff
@@ -3654,6 +3654,8 @@ refinable(?type(map, _) = Ty0, TEnv, Trace) ->
             %io:format("map ~p maybe refinable - checking assocs: ~p\n", [Ty, R]),
             length(Assocs) =/= 0 andalso R
     end;
+refinable(?type(string) = Ty, _TEnv, _Trace) ->
+    true;
```

Comment:
Good question, I have been wondering about the explanation myself for quite a while.

> I though it means that it's possible to refine a type to a smaller type by exhaustively covering part of the type [...]

Indeed, that's what comes to mind.

The comment at the top is helpful for a start:
```
% Is a type refinable to the point that we do exhaustiveness checking on it?
refinable(Ty, TEnv) ->
```

Then, we have to consider that `refinable` is recursive, so a type is only refinable if all the types it's made of are refinable.

@josefs original commit messages are also very insightful, especially these fragments:

> Here we define the set of types which can be checked
for exhaustiveness. -- from https://github.com/josefs/Gradualizer/pull/188/commits/4705a246ff0dc6525f7eb5b7e57589dce334e7c5

> This patch reuses the type refinement mechanism and checks that if we
the final refined types after having gone through all the clauses is
none(), then the clauses are exhaustive.

Unfortunately, the type refinement machinery is a little too coarse
since we must be able to refine to a sensible erlang type at each
intermediate step and that hinders us from capturing certain patterns
such as various forms of patterns on lists. Eventually, we will have
to implement a completely different machinery for exhaustiveness
checking to get something more precise, but this will do for now.

Implementing exhaustiveness checking revealed a bunch of imprecisions
in the type refinement code and I've fixed them in this patch. -- from https://github.com/josefs/Gradualizer/pull/188/commits/27ee3582a08b90902906f2f9607c669ccb6b7de0

The latter brings us to an example:

```
-type string_variant_t() :: {non_string, integer()}
                          | {string, string()}.

-spec string_variants(string_variant_t()) -> ok.
string_variants(T) ->
    case T of
        {string, _} -> ok
    end.
```

`T` has to be `refinable` to for Gradualizer to do exhaustiveness checking of the `case T of ... end`. This means all unions variants have to be refinable, which means that `string()` has to be refinable for the purpose of making the whole union `string_variant_t()` refinable, even though we cannot really refine `string()` in any meaningful way.

Ultimately, I think this stems from the fact that the original `patch reuses the type refinement mechanism`. I'm not sure if these mechanisms could be made completely separate, but maybe at least some code paths should be split between "actual refinement" and "exhaustiveness checking support" if we want to make it super clean. Maybe @josefs could chime in on that.

Reply:
I think `refinable` was originally a safeguard to protect from crashes on types known not to be supported. Well, let's see what happens when we just return `true` on everything :D

Reply:
Looking forward to see. I'm not a fan of safeguards. :-) It's like a workaround for something we can't solve the right way.

Reply:
Just for the record, the idea was tried out here - https://github.com/josefs/Gradualizer/pull/347 - and is not as straightforward as it seemed.

Reply:
That's true, but it tells us something about what matters for exhaustiveness checking. We can leave this to be solved in the future.